libs/gltfio/src/MorphHelper.cpp

```diff
@@ -155,13 +105,40 @@ void MorphHelper::addPrimitive(cgltf_mesh const* mesh, int primitiveIndex, Table
     const cgltf_primitive& prim = mesh->primitives[primitiveIndex];
     const auto& gltfioPrim = mAsset->mMeshCache.at(mesh)[primitiveIndex];
     VertexBuffer* vertexBuffer = gltfioPrim.vertices;
+    entry->primitives.push_back({ static_cast<int>(prim.targets_count) });
 
-    entry->primitives.push_back({ vertexBuffer });
     auto& morphHelperPrim = entry->primitives.back();
-
-    for (int i = 0; i < 4; i++) {
-        morphHelperPrim.positions[i] = gltfioPrim.morphPositions[i];
-        morphHelperPrim.tangents[i] = gltfioPrim.morphTangents[i];
+    auto& targets = morphHelperPrim.targets;
+
+    static auto getTargetsTextureSize = [](size_t vertexCount) -> int2 {
+        // Must equal to MAX_MORPH_TARGETS_WIDTH in getters.vs
+        constexpr size_t MAX_MORPH_TARGETS_WIDTH = 4096;
+        return { std::min(vertexCount, MAX_MORPH_TARGETS_WIDTH),
+                 ((vertexCount + MAX_MORPH_TARGETS_WIDTH) / MAX_MORPH_TARGETS_WIDTH) };
+    };
+
+    const auto targetsTextureSize = getTargetsTextureSize(vertexBuffer->getVertexCount());
+
+    // Compute the byte size to copy to the texture.
+    // It's required because when the vertex count exceeds the maximum texture width,
+    // we need to use the two-dim texture instead of the one-dim texture
+    // to avoid the hardware limitation.
+    // In this case, we need to allocate a memory which has the two-dim to copy properly.
+    const uint32_t requiredSize = Texture::PixelBufferDescriptor::computeDataSize(
+            Texture::PixelBufferDescriptor::PixelDataFormat::RGBA,
+            Texture::PixelBufferDescriptor::PixelDataType::FLOAT,
+            targetsTextureSize.x * sizeof(float4), targetsTextureSize.y, 1);
+
+    enum MorphTargetAttributes { POSITION = 0, NORMAL = 1, COUNT = 2 };
+
+    if (prim.targets_count) {
+        targets.targets = Texture::Builder()
+                .width(targetsTextureSize.x)
+                .height(targetsTextureSize.y)
+                .depth(prim.targets_count * MorphTargetAttributes::COUNT)
+                .format(Texture::InternalFormat::RGBA32F)
```

Comment:
RGBA32F seems overkill. Why not RGBA16F or even normalized integer formats?

Reply:
I try to optimize texture format. However the precision of RGBA16F or normalized integer formats is not enough for morph target position.